CIP-0031/README.md

```diff
@@ -0,0 +1,206 @@
+- Referenced outputs are _not_ removed from the UTXO set if the transaction validates.
+- Reference inputs _are_ visible to scripts.
+
+Finally, a transaction must _spend_ at least one output.[^2] 
+
+[^1]: Note in particular that this proposal does not affect _outputs_ in any way.
+[^2]: This restriction already exists, but repeat it here for clarity.
+
+### Script context
+
+Scripts are passed information about transactions via the script context.
+The script context therefore needs to be augmented to contain information about reference inputs.
+
+Changing the script context will require a new Plutus language version in the ledger to support the new interface.
+The change in the new interface is: a _new_ field is added to the structure which contains the list of reference inputs.
+
+Old versions of the language will retain the old interface.
+We do not propose to include information about reference inputs in the old interface.
+
+### Optional datums
+
+Today, transactions can contain optional datums that are not required for validation.
+Currently the optional datums must all be pre-images of datum hashes that appear in outputs.
+We change this so that the optional datums can also be pre-images of datum hashes that appear in reference inputs.[^3]
+
+Note that the existing mechanism includes the hash of the optional datum structure in the transaction body, so that additional datums cannot be stripped by an attacker without voiding transaction signatures.
+
+[^3]: Pre-images of datum hashes that appear in _spent_ inputs are of course mandatory.
+
+### CDDL
+
+The CDDL for transaction bodies will change to the following to reflect the new field.
+```
+transaction_body =
+ { 0 : set<transaction_input>    ; inputs
+ ...      
+ , ? 16 : set<transaction_input> ; reference inputs
+ }
+```
+
+## Rationale
+
+The key idea of this proposal is to use UTXOs to carry information. 
+But UTXOs are currently a bad fit for distributing information. 
+Because of locality, we have to include outputs that we use in the transaction, and the only way we have of doing that is to _spend_ them - and a spent output cannot then be referenced by anything else. 
+To put it another way: outputs are resource-like, but information is not resource-like.
+
+The solution is to add a way to _inspect_ ("reference") outputs without spending them. 
+This allows outputs to play double duty as resource containers (for the value they carry) and information containers (for the data they carry). 
+
+### Requirements
+
+We have a number of requirements that we need to fulfil.
+- Determinism
+    - It must be possible to predict the execution of scripts precisely, given the transaction.
+- Locality 
+    - All data involved in transaction validation should be included in the transaction or the outputs which it spends.
```

Comment:
Considering this proposal, should this be `or the previous outputs which it references.`

Reply:
Yes!